### ISSUE TITLE
extract subexpression does not replace all occurrences

### ISSUE THREAD

**Issue by Trass3r** [clangd/clangd#924]
```cpp
int foo()
{
    int x = 1 + 1;
    int y = 1 + 1;
    return x + y;
}
```
Extracting 1 + 1 leads to:
```cpp
    auto placeholder = 1 + 1;
    int x = placeholder;
    int y = 1 + 1;
```

The whole purpose of the transformation is not having to replace all the occurrences manually.

Also the cursor is at `placeholder^;` afterwards. Instead it should multi-cursor select all occurrences of the new name so it can be changed right away.

**Comment by HighCommander4** [comment]
My understanding is that this is not a goal of the "extract subexpression to variable" refactoring.

The goal is to take an expression (usually one which is **not** already the initializer of a variable, but rather a sub-expression of an initializer, or an expression used in some other context than an initializer), and extract a variable with that as its value, with the intent of you then subsequently adding new uses of that variable.

What you're suggesting here sounds like a new refactoring that's more challenging to implement, as it needs to search the code for other occurrences of the expression, **and** prove that the various occurrences of the expression do in fact have the same value (so, for example, if the expression references a non-const variable, clangd would need to prove that its value does not change between occurrences; if the expression includes a function call, clangd would need to prove that its return value does not change between calls). This latter check is data-dependent and not tractable in general (but could be in a subset of cases).

**Comment by Trass3r** [comment]
> usually one which is **not** already the complete initializer of a variable

Just a simplified example.

> it needs to search the code for other occurrences of the expression, **and** prove that the various occurrences of the expression do in fact have the same value

I don't expect it to be smart about it.
The reason for doing this kind of transformation is usually to get rid of copy-paste in existing code.
I just want it to automate the tedious steps.

**Comment by HighCommander4** [comment]
> I don't expect it to be smart about it.

If you don't expect it to be smart about it, then what would the refactoring give you over your text editor's search-and-replace?

**Comment by Trass3r** [comment]
By smart I meant it doesn't need to prove the expression always gives the same result. That's almost impossible anyway in general.
But it should restrict its actions to the current function/block. That's what the text editor doesn't do, it jumps around everywhere.
This nicely summarizes typical expectations:
https://docs.wholetomato.com/default.asp?W514

**Comment by HighCommander4** [comment]
Some text editors provide a "replace all in selection" feature that's handy for such occasions (though not all -- e.g. Eclipse has it, but I'm not seeing it in VSCode).

Eclipse also provides a shortcut for "select enclosing block", maybe we could add something like that to clangd.

It sounds like the desired action could be achieved using those + textual replace.

**Comment by kadircet** [comment]
I agree with what Nathan mentioned above. In general it's not feasible for clangd to prove that two same expressions from different contexts evaluate to the same value.

Hence this will restrict the feature only to the constant expressions and having multiple occurences of those is IMO quite rare (especially in the same code block, not even in the same file. you mention a copy paste approach, but is it really common for you to copy paste a sub expression into multiple places inside the **same code block**  or do you just put it into a variable and copy paste its name instead?)

As for cursor position, LSP doesn't provide a way for a server to say "put the cursor into this place" let alone multiple places after execution of a code action. So this needs to be implemented on client/editor side or the protocol needs to be changed. You can issue a rename request immediately afterwards though.

**Comment by Trass3r** [comment]
> Hence this will restrict the feature only to the constant expressions and having multiple occurences of those is IMO quite rare 

There are pure and idempotent function calls for example. For the latter we don't even have an [attribute](https://bugs.llvm.org/show_bug.cgi?id=34600). The former is rarely used. Clangd can't prove anything in general and doesn't need to imho. It's just about automation.

> you mention a copy paste approach, but is it really common for you to copy paste a sub expression into multiple places inside the **same code block** or do you just put it into a variable and copy paste its name instead?

I've seen people copy-paste whole functions which were already way too large just to change 1 or 2 occurrences of something instead of introducing another parameter. There's really no limit to this insanity in the world.

> As for cursor position, LSP doesn't provide a way for a server to say "put the cursor into this place" let alone multiple places after execution of a code action. So this needs to be implemented on client/editor side or the protocol needs to be changed. You can issue a rename request immediately afterwards though.

Yeah I kind of already anticipated that. But would be a cool feature.
